--- a/vendor/psm-0.1.26/src/arch/zseries_linux.s	2026-01-19 16:34:28.000000000 +0000
+++ b/vendor/psm-0.1.26/src/arch/zseries_linux.s	2026-03-19 08:25:34.977512753 +0000
@@ -46,7 +46,8 @@
 /* extern "C" fn(r2: usize, r3: extern "C" fn(usize), r4: *mut u8) */
 .cfi_startproc
     /* FIXME: backtrace does not terminate cleanly for some reason */
-    lay %r15, -160(%r4)
+    lgr %r15, %r4
+    aghi %r15, -160
     /* FIXME: this is `basr` instead of `br` purely to remove the backtrace link to the caller */
     basr %r14, %r3
 .rust_psm_replace_stack_end:
@@ -60,8 +61,10 @@
 rust_psm_on_stack:
 /* extern "C" fn(r2: usize, r3: usize, r4: extern "C" fn(usize, usize), r5: *mut u8) */
 .cfi_startproc
-    stmg %r14, %r15, -16(%r5)
-    lay %r15, -176(%r5)
+    aghi %r5, -16
+    stmg %r14, %r15, 0(%r5)
+    lgr %r15, %r5
+    aghi %r15, -160
     .cfi_def_cfa %r15, 176
     .cfi_offset %r14, -16
     .cfi_offset %r15, -8
