diff -ruN a/include/llvm/BinaryFormat/ELF.h b/include/llvm/BinaryFormat/ELF.h
--- a/include/llvm/BinaryFormat/ELF.h	2026-07-07 15:50:02.244521783 -0400
+++ b/include/llvm/BinaryFormat/ELF.h	2026-07-07 15:50:02.246903594 -0400
@@ -994,6 +994,11 @@
 #include "ELFRelocs/M68k.def"
 };
 
+// ELF Relocation types for TILE-Gx
+enum {
+#include "ELFRelocs/TileGX.def"
+};
+
 // MSP430 specific e_flags
 enum : unsigned {
   EF_MSP430_MACH_MSP430x11 = 11,
diff -ruN a/include/llvm/BinaryFormat/ELFRelocs/TileGX.def b/include/llvm/BinaryFormat/ELFRelocs/TileGX.def
--- a/include/llvm/BinaryFormat/ELFRelocs/TileGX.def	1969-12-31 19:00:00.000000000 -0500
+++ b/include/llvm/BinaryFormat/ELFRelocs/TileGX.def	2026-07-07 15:50:02.254049100 -0400
@@ -0,0 +1,125 @@
+
+#ifndef ELF_RELOC
+#error "ELF_RELOC must be defined"
+#endif
+
+ELF_RELOC(R_TILEGX_NONE,                          0)
+ELF_RELOC(R_TILEGX_64,                            1)
+ELF_RELOC(R_TILEGX_32,                            2)
+ELF_RELOC(R_TILEGX_16,                            3)
+ELF_RELOC(R_TILEGX_8,                             4)
+ELF_RELOC(R_TILEGX_64_PCREL,                      5)
+ELF_RELOC(R_TILEGX_32_PCREL,                      6)
+ELF_RELOC(R_TILEGX_16_PCREL,                      7)
+ELF_RELOC(R_TILEGX_8_PCREL,                       8)
+ELF_RELOC(R_TILEGX_HW0,                           9)
+ELF_RELOC(R_TILEGX_HW1,                          10)
+ELF_RELOC(R_TILEGX_HW2,                          11)
+ELF_RELOC(R_TILEGX_HW3,                          12)
+ELF_RELOC(R_TILEGX_HW0_LAST,                     13)
+ELF_RELOC(R_TILEGX_HW1_LAST,                     14)
+ELF_RELOC(R_TILEGX_HW2_LAST,                     15)
+ELF_RELOC(R_TILEGX_COPY,                         16)
+ELF_RELOC(R_TILEGX_GLOB_DAT,                     17)
+ELF_RELOC(R_TILEGX_JMP_SLOT,                     18)
+ELF_RELOC(R_TILEGX_RELATIVE,                     19)
+ELF_RELOC(R_TILEGX_BROFF_X1,                     20)
+ELF_RELOC(R_TILEGX_JUMPOFF_X1,                   21)
+ELF_RELOC(R_TILEGX_JUMPOFF_X1_PLT,               22)
+ELF_RELOC(R_TILEGX_IMM8_X0,                      23)
+ELF_RELOC(R_TILEGX_IMM8_Y0,                      24)
+ELF_RELOC(R_TILEGX_IMM8_X1,                      25)
+ELF_RELOC(R_TILEGX_IMM8_Y1,                      26)
+ELF_RELOC(R_TILEGX_DEST_IMM8_X1,                 27)
+ELF_RELOC(R_TILEGX_MT_IMM14_X1,                  28)
+ELF_RELOC(R_TILEGX_MF_IMM14_X1,                  29)
+ELF_RELOC(R_TILEGX_MMSTART_X0,                   30)
+ELF_RELOC(R_TILEGX_MMEND_X0,                     31)
+ELF_RELOC(R_TILEGX_SHAMT_X0,                     32)
+ELF_RELOC(R_TILEGX_SHAMT_X1,                     33)
+ELF_RELOC(R_TILEGX_SHAMT_Y0,                     34)
+ELF_RELOC(R_TILEGX_SHAMT_Y1,                     35)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0,                 36)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0,                 37)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW1,                 38)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW1,                 39)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW2,                 40)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW2,                 41)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW3,                 42)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW3,                 43)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0_LAST,            44)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0_LAST,            45)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW1_LAST,            46)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW1_LAST,            47)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW2_LAST,            48)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW2_LAST,            49)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0_PCREL,           50)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0_PCREL,           51)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW1_PCREL,           52)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW1_PCREL,           53)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW2_PCREL,           54)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW2_PCREL,           55)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW3_PCREL,           56)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW3_PCREL,           57)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0_LAST_PCREL,      58)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0_LAST_PCREL,      59)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW1_LAST_PCREL,      60)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW1_LAST_PCREL,      61)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW2_LAST_PCREL,      62)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW2_LAST_PCREL,      63)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0_GOT,             64)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0_GOT,             65)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0_PLT_PCREL,       66)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0_PLT_PCREL,       67)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW1_PLT_PCREL,       68)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW1_PLT_PCREL,       69)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW2_PLT_PCREL,       70)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW2_PLT_PCREL,       71)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0_LAST_GOT,        72)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0_LAST_GOT,        73)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW1_LAST_GOT,        74)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW1_LAST_GOT,        75)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW3_PLT_PCREL,       76)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW3_PLT_PCREL,       77)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0_TLS_GD,          78)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0_TLS_GD,          79)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0_TLS_LE,          80)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0_TLS_LE,          81)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0_LAST_TLS_LE,     82)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0_LAST_TLS_LE,     83)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW1_LAST_TLS_LE,     84)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW1_LAST_TLS_LE,     85)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0_LAST_TLS_GD,     86)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0_LAST_TLS_GD,     87)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW1_LAST_TLS_GD,     88)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW1_LAST_TLS_GD,     89)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0_TLS_IE,          92)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0_TLS_IE,          93)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0_LAST_PLT_PCREL,  94)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0_LAST_PLT_PCREL,  95)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW1_LAST_PLT_PCREL,  96)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW1_LAST_PLT_PCREL,  97)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW2_LAST_PLT_PCREL,  98)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW2_LAST_PLT_PCREL,  99)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW0_LAST_TLS_IE,    100)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW0_LAST_TLS_IE,    101)
+ELF_RELOC(R_TILEGX_IMM16_X0_HW1_LAST_TLS_IE,    102)
+ELF_RELOC(R_TILEGX_IMM16_X1_HW1_LAST_TLS_IE,    103)
+ELF_RELOC(R_TILEGX_TLS_DTPMOD64,                106)
+ELF_RELOC(R_TILEGX_TLS_DTPOFF64,                107)
+ELF_RELOC(R_TILEGX_TLS_TPOFF64,                 108)
+ELF_RELOC(R_TILEGX_TLS_DTPMOD32,                109)
+ELF_RELOC(R_TILEGX_TLS_DTPOFF32,                110)
+ELF_RELOC(R_TILEGX_TLS_TPOFF32,                 111)
+ELF_RELOC(R_TILEGX_TLS_GD_CALL,                 112)
+ELF_RELOC(R_TILEGX_IMM8_X0_TLS_GD_ADD,          113)
+ELF_RELOC(R_TILEGX_IMM8_X1_TLS_GD_ADD,          114)
+ELF_RELOC(R_TILEGX_IMM8_Y0_TLS_GD_ADD,          115)
+ELF_RELOC(R_TILEGX_IMM8_Y1_TLS_GD_ADD,          116)
+ELF_RELOC(R_TILEGX_TLS_IE_LOAD,                 117)
+ELF_RELOC(R_TILEGX_IMM8_X0_TLS_ADD,             118)
+ELF_RELOC(R_TILEGX_IMM8_X1_TLS_ADD,             119)
+ELF_RELOC(R_TILEGX_IMM8_Y0_TLS_ADD,             120)
+ELF_RELOC(R_TILEGX_IMM8_Y1_TLS_ADD,             121)
+ELF_RELOC(R_TILEGX_GNU_VTINHERIT,               128)
+ELF_RELOC(R_TILEGX_GNU_VTENTRY,                 129)
diff -ruN a/include/llvm/IR/CMakeLists.txt b/include/llvm/IR/CMakeLists.txt
--- a/include/llvm/IR/CMakeLists.txt	2026-07-07 15:50:02.261401837 -0400
+++ b/include/llvm/IR/CMakeLists.txt	2026-07-07 15:50:02.263964506 -0400
@@ -25,4 +25,5 @@
 tablegen(LLVM IntrinsicsX86.h -gen-intrinsic-enums -intrinsic-prefix=x86)
 tablegen(LLVM IntrinsicsXCore.h -gen-intrinsic-enums -intrinsic-prefix=xcore)
 tablegen(LLVM IntrinsicsVE.h -gen-intrinsic-enums -intrinsic-prefix=ve)
+tablegen(LLVM IntrinsicsTile.h -gen-intrinsic-enums -intrinsic-prefix=tilegx)
 add_public_tablegen_target(intrinsics_gen)
diff -ruN a/include/llvm/IR/Intrinsics.td b/include/llvm/IR/Intrinsics.td
--- a/include/llvm/IR/Intrinsics.td	2026-07-07 15:50:02.271065386 -0400
+++ b/include/llvm/IR/Intrinsics.td	2026-07-07 15:50:02.274049592 -0400
@@ -2871,5 +2871,6 @@
 include "llvm/IR/IntrinsicsVE.td"
 include "llvm/IR/IntrinsicsDirectX.td"
 include "llvm/IR/IntrinsicsLoongArch.td"
+include "llvm/IR/IntrinsicsTile.td"
 
 #endif // TEST_INTRINSICS_SUPPRESS_DEFS
diff -ruN a/include/llvm/IR/IntrinsicsTile.td b/include/llvm/IR/IntrinsicsTile.td
--- a/include/llvm/IR/IntrinsicsTile.td	1969-12-31 19:00:00.000000000 -0500
+++ b/include/llvm/IR/IntrinsicsTile.td	2026-07-07 15:50:02.281049765 -0400
@@ -0,0 +1,399 @@
+//===- IntrinsicsTile.td - Defines TILE-Gx intrinsics ---------*- tablegen -*-===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines all of the TILE-Gx intrinsics.
+//
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+
+let TargetPrefix = "tilegx" in {  // All intrinsics start with "llvm.tilegx.".
+
+// v1XXX
+def int_tilegx_v1add : ClangBuiltin<"__insn_v1add">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], [Commutative]>;
+
+def int_tilegx_v1addi : ClangBuiltin<"__insn_v1addi">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1adduc : ClangBuiltin<"__insn_v1adduc">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1adiffu : ClangBuiltin<"__insn_v1adiffu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1avgu : ClangBuiltin<"__insn_v1avgu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1cmpeq : ClangBuiltin<"__insn_v1cmpeq">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1cmpeqi : ClangBuiltin<"__insn_v1cmpeqi">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1cmples : ClangBuiltin<"__insn_v1cmples">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1cmpleu : ClangBuiltin<"__insn_v1cmpleu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1cmplts : ClangBuiltin<"__insn_v1cmplts">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1cmpltsi : ClangBuiltin<"__insn_v1cmpltsi">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1cmpltu : ClangBuiltin<"__insn_v1cmpltu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1cmpltui : ClangBuiltin<"__insn_v1cmpltui">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1cmpne : ClangBuiltin<"__insn_v1cmpne">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1ddotpu : ClangBuiltin<"__insn_v1ddotpu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1ddotpua : ClangBuiltin<"__insn_v1ddotpua">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1ddotpus : ClangBuiltin<"__insn_v1ddotpus">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1ddotpusa : ClangBuiltin<"__insn_v1ddotpusa">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1dotp : ClangBuiltin<"__insn_v1dotp">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1dotpa : ClangBuiltin<"__insn_v1dotpa">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1dotpu : ClangBuiltin<"__insn_v1dotpu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1dotpua : ClangBuiltin<"__insn_v1dotpua">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1dotpus : ClangBuiltin<"__insn_v1dotpus">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1dotpusa : ClangBuiltin<"__insn_v1dotpusa">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1int_h : ClangBuiltin<"__insn_v1int_h">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1int_l : ClangBuiltin<"__insn_v1int_l">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1maxu : ClangBuiltin<"__insn_v1maxu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1maxui : ClangBuiltin<"__insn_v1maxui">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1minu : ClangBuiltin<"__insn_v1minu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1minui : ClangBuiltin<"__insn_v1minui">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1mnz : ClangBuiltin<"__insn_v1mnz">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1multu : ClangBuiltin<"__insn_v1multu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1mulu : ClangBuiltin<"__insn_v1mulu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1mulus : ClangBuiltin<"__insn_v1mulus">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1mz : ClangBuiltin<"__insn_v1mz">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1sadau : ClangBuiltin<"__insn_v1sadau">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1sadu : ClangBuiltin<"__insn_v1sadu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1shl : ClangBuiltin<"__insn_v1shl">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1shli : ClangBuiltin<"__insn_v1shli">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1shrs : ClangBuiltin<"__insn_v1shrs">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1shrsi : ClangBuiltin<"__insn_v1shrsi">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1shru : ClangBuiltin<"__insn_v1shru">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1shrui : ClangBuiltin<"__insn_v1shrui">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1sub : ClangBuiltin<"__insn_v1sub">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v1subuc : ClangBuiltin<"__insn_v1subuc">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+// v2XXX
+def int_tilegx_v2add : ClangBuiltin<"__insn_v2add">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], [Commutative]>;
+
+def int_tilegx_v2addi : ClangBuiltin<"__insn_v2addi">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2addsc : ClangBuiltin<"__insn_v2addsc">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2adiffs : ClangBuiltin<"__insn_v2adiffs">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2avgs : ClangBuiltin<"__insn_v2avgs">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2cmpeq : ClangBuiltin<"__insn_v2cmpeq">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2cmpeqi : ClangBuiltin<"__insn_v2cmpeqi">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2cmples : ClangBuiltin<"__insn_v2cmples">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2cmpleu : ClangBuiltin<"__insn_v2cmpleu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2cmplts : ClangBuiltin<"__insn_v2cmplts">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2cmpltsi : ClangBuiltin<"__insn_v2cmpltsi">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2cmpltu : ClangBuiltin<"__insn_v2cmpltu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2cmpltui : ClangBuiltin<"__insn_v2cmpltui">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2cmpne : ClangBuiltin<"__insn_v2cmpne">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2dotp : ClangBuiltin<"__insn_v2dotp">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2dotpa : ClangBuiltin<"__insn_v2dotpa">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2int_h : ClangBuiltin<"__insn_v2int_h">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2int_l : ClangBuiltin<"__insn_v2int_l">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2maxs : ClangBuiltin<"__insn_v2maxs">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2maxsi : ClangBuiltin<"__insn_v2maxsi">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2mins : ClangBuiltin<"__insn_v2mins">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2minsi : ClangBuiltin<"__insn_v2minsi">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2mnz : ClangBuiltin<"__insn_v2mnz">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2mulfsc : ClangBuiltin<"__insn_v2mulfsc">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2muls : ClangBuiltin<"__insn_v2muls">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2mults : ClangBuiltin<"__insn_v2mults">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2mz : ClangBuiltin<"__insn_v2mz">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2packh : ClangBuiltin<"__insn_v2packh">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2packl : ClangBuiltin<"__insn_v2packl">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2packuc : ClangBuiltin<"__insn_v2packuc">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2sadas : ClangBuiltin<"__insn_v2sadas">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2sadau : ClangBuiltin<"__insn_v2sadau">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2sads : ClangBuiltin<"__insn_v2sads">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2sadu : ClangBuiltin<"__insn_v2sadu">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2shl : ClangBuiltin<"__insn_v2shl">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2shli : ClangBuiltin<"__insn_v2shli">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2shlsc : ClangBuiltin<"__insn_v2shlsc">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2shrs : ClangBuiltin<"__insn_v2shrs">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2shrsi : ClangBuiltin<"__insn_v2shrsi">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2shru : ClangBuiltin<"__insn_v2shru">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2shrui : ClangBuiltin<"__insn_v2shrui">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2sub : ClangBuiltin<"__insn_v2sub">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v2subsc : ClangBuiltin<"__insn_v2subsc">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+// v4XXX
+def int_tilegx_v4add : ClangBuiltin<"__insn_v4add">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], [Commutative]>;
+
+def int_tilegx_v4addsc : ClangBuiltin<"__insn_v4addsc">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v4int_h : ClangBuiltin<"__insn_v4int_h">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v4int_l : ClangBuiltin<"__insn_v4int_l">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v4packsc : ClangBuiltin<"__insn_v4packsc">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v4shl : ClangBuiltin<"__insn_v4shl">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v4shlsc : ClangBuiltin<"__insn_v4shlsc">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v4shrs : ClangBuiltin<"__insn_v4shrs">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v4shru : ClangBuiltin<"__insn_v4shru">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v4sub : ClangBuiltin<"__insn_v4sub">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+def int_tilegx_v4subsc : ClangBuiltin<"__insn_v4subsc">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], []>;
+
+// Network
+def int_tilegx_idn0r : ClangBuiltin<"__tile_idn0_receive">,
+  Intrinsic<[llvm_i64_ty], [], []>;
+
+def int_tilegx_idn1r : ClangBuiltin<"__tile_idn1_receive">,
+  Intrinsic<[llvm_i64_ty], [], []>;
+
+def int_tilegx_udn0r : ClangBuiltin<"__tile_udn0_receive">,
+  Intrinsic<[llvm_i64_ty], [], []>;
+
+def int_tilegx_udn1r : ClangBuiltin<"__tile_udn1_receive">,
+  Intrinsic<[llvm_i64_ty], [], []>;
+
+def int_tilegx_udn2r : ClangBuiltin<"__tile_udn2_receive">,
+  Intrinsic<[llvm_i64_ty], [], []>;
+
+def int_tilegx_udn3r : ClangBuiltin<"__tile_udn3_receive">,
+  Intrinsic<[llvm_i64_ty], [], []>;
+
+def int_tilegx_netbarrier : ClangBuiltin<"__tile_network_barrier">,
+  Intrinsic<[], [], []>;
+
+def int_tilegx_usend : ClangBuiltin<"__tile_udn_send">,
+  Intrinsic<[], [llvm_i64_ty], []>;
+
+def int_tilegx_isend : ClangBuiltin<"__tile_idn_send">,
+  Intrinsic<[], [llvm_i64_ty], []>;
+
+// Others
+def int_tilegx_mfspr : ClangBuiltin<"__insn_mfspr">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty], []>;
+
+def int_tilegx_mtspr : ClangBuiltin<"__insn_mtspr">,
+  Intrinsic<[], [llvm_i64_ty, llvm_i64_ty], []>;
+
+// Scalar arithmetic and bit manipulation
+def int_tilegx_addx : ClangBuiltin<"__insn_addx">,
+  Intrinsic<[llvm_i32_ty], [llvm_i32_ty, llvm_i32_ty], [IntrNoMem, Commutative]>;
+def int_tilegx_clz : ClangBuiltin<"__insn_clz">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty], [IntrNoMem]>;
+def int_tilegx_ctz : ClangBuiltin<"__insn_ctz">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty], [IntrNoMem]>;
+def int_tilegx_pcnt : ClangBuiltin<"__insn_pcnt">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty], [IntrNoMem]>;
+def int_tilegx_revbytes : ClangBuiltin<"__insn_revbytes">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty], [IntrNoMem]>;
+def int_tilegx_revbits : ClangBuiltin<"__insn_revbits">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty], [IntrNoMem]>;
+def int_tilegx_crc32_32 : ClangBuiltin<"__insn_crc32_32">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], [IntrNoMem]>;
+def int_tilegx_crc32_8 : ClangBuiltin<"__insn_crc32_8">,
+  Intrinsic<[llvm_i64_ty], [llvm_i64_ty, llvm_i64_ty], [IntrNoMem]>;
+def int_tilegx_mm : ClangBuiltin<"__insn_mm">,
+  Intrinsic<[llvm_i64_ty],
+            [llvm_i64_ty, llvm_i64_ty, llvm_i64_ty, llvm_i64_ty], [IntrNoMem]>;
+
+// Atomic memory operations. The first argument is the address.
+def int_tilegx_exch : ClangBuiltin<"__insn_exch">,
+  Intrinsic<[llvm_i64_ty], [llvm_ptr_ty, llvm_i64_ty], [IntrArgMemOnly]>;
+def int_tilegx_exch4 : ClangBuiltin<"__insn_exch4">,
+  Intrinsic<[llvm_i32_ty], [llvm_ptr_ty, llvm_i32_ty], [IntrArgMemOnly]>;
+def int_tilegx_fetchadd : ClangBuiltin<"__insn_fetchadd">,
+  Intrinsic<[llvm_i64_ty], [llvm_ptr_ty, llvm_i64_ty], [IntrArgMemOnly]>;
+def int_tilegx_fetchadd4 : ClangBuiltin<"__insn_fetchadd4">,
+  Intrinsic<[llvm_i32_ty], [llvm_ptr_ty, llvm_i32_ty], [IntrArgMemOnly]>;
+def int_tilegx_fetchand : ClangBuiltin<"__insn_fetchand">,
+  Intrinsic<[llvm_i64_ty], [llvm_ptr_ty, llvm_i64_ty], [IntrArgMemOnly]>;
+def int_tilegx_fetchand4 : ClangBuiltin<"__insn_fetchand4">,
+  Intrinsic<[llvm_i32_ty], [llvm_ptr_ty, llvm_i32_ty], [IntrArgMemOnly]>;
+def int_tilegx_fetchor : ClangBuiltin<"__insn_fetchor">,
+  Intrinsic<[llvm_i64_ty], [llvm_ptr_ty, llvm_i64_ty], [IntrArgMemOnly]>;
+def int_tilegx_fetchor4 : ClangBuiltin<"__insn_fetchor4">,
+  Intrinsic<[llvm_i32_ty], [llvm_ptr_ty, llvm_i32_ty], [IntrArgMemOnly]>;
+def int_tilegx_fetchaddgez : ClangBuiltin<"__insn_fetchaddgez">,
+  Intrinsic<[llvm_i64_ty], [llvm_ptr_ty, llvm_i64_ty], [IntrArgMemOnly]>;
+def int_tilegx_fetchaddgez4 : ClangBuiltin<"__insn_fetchaddgez4">,
+  Intrinsic<[llvm_i32_ty], [llvm_ptr_ty, llvm_i32_ty], [IntrArgMemOnly]>;
+def int_tilegx_cmpexch : ClangBuiltin<"__insn_cmpexch">,
+  Intrinsic<[llvm_i64_ty], [llvm_ptr_ty, llvm_i64_ty], [IntrArgMemOnly]>;
+def int_tilegx_cmpexch4 : ClangBuiltin<"__insn_cmpexch4">,
+  Intrinsic<[llvm_i32_ty], [llvm_ptr_ty, llvm_i32_ty], [IntrArgMemOnly]>;
+} // end TargetPrefix
diff -ruN a/include/llvm/TargetParser/Triple.h b/include/llvm/TargetParser/Triple.h
--- a/include/llvm/TargetParser/Triple.h	2026-07-07 15:50:02.287564523 -0400
+++ b/include/llvm/TargetParser/Triple.h	2026-07-07 15:50:02.289636589 -0400
@@ -83,6 +83,7 @@
     systemz,        // SystemZ: s390x
     tce,            // TCE (http://tce.cs.tut.fi/): tce
     tcele,          // TCE little endian (http://tce.cs.tut.fi/): tcele
+    tilegx,         // TILE-Gx (64-bit, little endian): tilegx
     thumb,          // Thumb (little endian): thumb, thumbv.*
     thumbeb,        // Thumb (big endian): thumbeb
     x86,            // X86: i[3-9]86
diff -ruN a/lib/IR/Intrinsics.cpp b/lib/IR/Intrinsics.cpp
--- a/lib/IR/Intrinsics.cpp	2026-07-07 15:50:02.295846197 -0400
+++ b/lib/IR/Intrinsics.cpp	2026-07-07 15:50:02.298113563 -0400
@@ -28,6 +28,7 @@
 #include "llvm/IR/IntrinsicsRISCV.h"
 #include "llvm/IR/IntrinsicsS390.h"
 #include "llvm/IR/IntrinsicsSPIRV.h"
+#include "llvm/IR/IntrinsicsTile.h"
 #include "llvm/IR/IntrinsicsVE.h"
 #include "llvm/IR/IntrinsicsX86.h"
 #include "llvm/IR/IntrinsicsXCore.h"
diff -ruN a/lib/Object/ELF.cpp b/lib/Object/ELF.cpp
--- a/lib/Object/ELF.cpp	2026-07-07 15:50:02.304050330 -0400
+++ b/lib/Object/ELF.cpp	2026-07-07 15:50:02.306050379 -0400
@@ -183,6 +183,13 @@
       break;
     }
     break;
+  case ELF::EM_TILEGX:
+    switch (Type) {
+#include "llvm/BinaryFormat/ELFRelocs/TileGX.def"
+    default:
+      break;
+    }
+    break;
   default:
     break;
   }
diff -ruN a/lib/TargetParser/Triple.cpp b/lib/TargetParser/Triple.cpp
--- a/lib/TargetParser/Triple.cpp	2026-07-07 15:50:02.312050527 -0400
+++ b/lib/TargetParser/Triple.cpp	2026-07-07 15:50:02.313909097 -0400
@@ -76,6 +76,7 @@
   case systemz:        return "s390x";
   case tce:            return "tce";
   case tcele:          return "tcele";
+  case tilegx:         return "tilegx";
   case thumb:          return "thumb";
   case thumbeb:        return "thumbeb";
   case ve:             return "ve";
@@ -249,6 +250,8 @@
   case dxil:        return "dx";
 
   case xtensa:      return "xtensa";
+
+  case tilegx:      return "tilegx";
   }
 }
 
@@ -461,6 +464,7 @@
     .Case("systemz", systemz)
     .Case("tce", tce)
     .Case("tcele", tcele)
+    .Case("tilegx", tilegx)
     .Case("thumb", thumb)
     .Case("thumbeb", thumbeb)
     .Case("x86", x86)
@@ -606,6 +610,7 @@
           .Cases("sparcv9", "sparc64", Triple::sparcv9)
           .Case("tce", Triple::tce)
           .Case("tcele", Triple::tcele)
+          .Case("tilegx", Triple::tilegx)
           .Case("xcore", Triple::xcore)
           .Case("nvptx", Triple::nvptx)
           .Case("nvptx64", Triple::nvptx64)
@@ -976,6 +981,7 @@
   case Triple::spir:
   case Triple::tce:
   case Triple::tcele:
+  case Triple::tilegx:
   case Triple::thumbeb:
   case Triple::ve:
   case Triple::xcore:
@@ -1725,6 +1731,7 @@
   case llvm::Triple::spir64:
   case llvm::Triple::spirv64:
   case llvm::Triple::systemz:
+  case llvm::Triple::tilegx:
   case llvm::Triple::ve:
   case llvm::Triple::wasm64:
   case llvm::Triple::x86_64:
@@ -2062,6 +2069,7 @@
   case Triple::spirv32:
   case Triple::spirv64:
   case Triple::tcele:
+  case Triple::tilegx:
   case Triple::thumb:
   case Triple::ve:
   case Triple::wasm32:
@@ -2280,7 +2288,7 @@
   }
 
   if (isAArch64() || isX86() || isPPC() || isMIPS() || isSPARC() || isBPF() ||
-      isRISCV() || isLoongArch())
+      isRISCV() || isLoongArch() || getArch() == Triple::tilegx)
     return ExceptionHandling::DwarfCFI;
 
   switch (getArch()) {
